### PR TITLE
[Proposal] Pool WriteContext per thread

### LINKED ISSUE #165: Pool UvWriteReq objects
It might help.

### LINKED ISSUE #289: [Perf] High Allocator DoWriteIfNeeded
Drill down on Issue #288
Measurements post PR #287

Item 5 on highest allocators in plaintext test

most-memory

Allocations (2003 requests)

alloc

alloc

Writes alloc on List.Add not `GCHandle.Alloc` for clarity

Bytes (2003 requests)

bytes

bytes

Writes alloc on List.Add not `GCHandle.Alloc` for clarity

### LINKED ISSUE #290: [Perf] High Allocator SocketOutput.Write
Drill down on Issue #288
Measurements post PR #287

Item 2 on highest allocators in plaintext test

most-memory

`Microsoft.AspNet.Server.Kestrel.Http.SocketOutput.Write`

Allocations (2003 requests)

allocations

Bytes (2003 requests)

bytes

### PR DESCRIPTION
**Pull request by benaadams**
Continuation of SocketOutput Allocation Reductions #466

With extra of "Pool WriteContext per thread rather than connection"

Resolves #165 Pool UvWriteReq objects
Resolves #289 High Allocator DoWriteIfNeeded
Resolves #290 High Allocator SocketOutput.Write

### CODE REVIEW

**Comment by benaadams** [comment]
Need to look into this a bit more, weird things happening.

**Comment by benaadams** [comment]
Wider pooling obv needed Self resetting...

File: src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs [hunk 1]
```diff
@@ -14,12 +14,18 @@ namespace Microsoft.AspNet.Server.Kestrel.Http
...
+        public const int MaxPooledWriteReqs = 1024;
+
         private const int _maxPendingWrites = 3;
         private const int _maxBytesPreCompleted = 65536;
         private const int _initialTaskQueues = 64;
+        private const int _maxWriteContextsPooledPerThread = 128;
 
         private static WaitCallback _returnBlocks = (state) => ReturnBlocks((MemoryPoolBlock2)state);
 
+        [ThreadStatic]
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs]: I'm afraid of adding a ThreadStatic. We don't use ThreadStatics anywhere else in Kestrel that I'm aware of.

I just checked the default maximum number of worker threads on my machine, and it was 32,767. I don't like having potentially that many static queues. Considering that each queue could pool up to 128 WriteContexts (and you just calculated the size WriteContext being 88 bytes), that's up to ~370MB of WriteContexts that will never be freed. That's never even considering the size of the queues themselves.

**Comment by halter73** [comment]
There's nothing in here that I need if I just want to merge #466, right?

**Comment by benaadams** [comment]
> There's nothing in here that I need if I just want to merge #466, right?

@halter73 Nope. #466 is good to merge.

File: src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs (hunk 1 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs]: Don't hold off merging #466 based on this (no issues in that one; and this is straight on top)

Know what you mean about `ThreadStatic`s though been doing some investigation. 

So the `ThreadStatic` is GC'd after thread exit; assuming it has no other outstanding strong references - which should be ok for this. Objects with a finalizer aren't so good as they'll hang around till the finalizer runs - but the WriteContext that's been reset is just a POCO managed type; so the queue will be happily GC'd when the threadpool scales back down.

If you are up to 32k of live threadpool threads; then the 32GB of thread stack space they need might be more of an issue than the 370MB of pooled `WriteContext`s.

However... I'm not sure this PR is a great win; it should reduce the memory for a very large number of connections - but that scenerio will probably have all sorts of other stuff to look at.

Happy to close for now and come back to; if it is an issue?

**Comment by benaadams** [comment]
Closing this; can revisit later

**Pull request closed without merging**